### PR TITLE
refactor: move methods under CChainState (pt. 1)

### PR DESCRIPTION
**Pull request by jamesob**
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11):

Parent PR: #15606 
Issue: #15605 
Specification: https://github.com/jamesob/assumeutxo-docs/tree/2019-04-proposal/proposal

---

This changeset starts moving functionality intimately related to CChainState into methods. Parameterizing these functions by a particular CChainState is necessary for the use of multiple chainstates simultaneously (e.g. for asynchronous background validation).

In this change, we
- make the CChainState interface public - since other units will start to invoke its methods directly,
- introduce `::ChainstateActive()`, the CChainState equivalent for `::ChainActive()`,
- and move `IsInitialBlockDownload()` and `FlushStateToDisk()` into methods on CChainState.

Independent of assumeutxo, these changes better encapsulate chainstate behavior and allow easier use from a testing context.

There are more methods that we'll move in the future, but they require other substantial changes (i.e. moving ownership of the `CCoinsView*` hierarchy into CChainState) so we'll save them for future PRs.

--- 

The first move-only commit is most easily reviewed with `git diff ... --color-moved=dimmed_zebra`.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -34,7 +35,6 @@
 #include <timedata.h>
 #include <tinyformat.h>
 #include <txdb.h>
-#include <txmempool.h>
```
Comment (MarcoFalke) [inline src/validation.cpp]: style-nit: Why shuffle the includes for no reason?

Comment (jamesob) [reply src/validation.cpp]: Oops, will fix.

**Comment by MarcoFalke** [comment]
I dislike that member functions that previously could (and should) only be called in the validation module are now callable from anywhere. To make it worse, they are easily confused e.g. `::ActivateBestChain` vs `CChainState::ActivateBestChain`. The only solution I could think of is to mark the public methods that are not supposed to be called outside of validation with an underscore or something (similar to python).

**Comment by MarcoFalke** [comment]
An alternative could be to remove the global methods (e.g. `::ActivateBestChain`) and inline them into the members?

**Review by Sjors — COMMENTED** [review]
Concept ACK

Can confirm 802ae945b4b0e87be78ab84e2954e69a4abc2f22 is move-only, but see inline comment about `CBlockIndexWorkComparator`.

I did an IBD on testnet with 65e2268, stopping it a few times (which I assume covers both functions you moved) and checking `getblockchaininfo`.

File: src/validation.h [hunk 2]
```diff
@@ -422,6 +422,195 @@ inline CBlockIndex* LookupBlockIndex(const uint256& hash)
...
+enum class FlushStateMode {
+    NONE,
+    IF_NEEDED,
+    PERIODIC,
+    ALWAYS
+};
+
+struct CBlockIndexWorkComparator
+{
+    bool operator()(const CBlockIndex *pa, const CBlockIndex *pb) const {
```
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/validation.h]: The method implementation doesn't, but the struct declaration does because it affects the layout of CChainState objects.

**Comment by Sjors** [comment]
> could (and should) only be called in the validation module are now callable from anywhere

Would you say the root cause of that is that ::ChainstateActive is global? Maybe we can make that only available through interfaces? There's not that many places we access member functions from.

**Comment by MarcoFalke** [comment]
> Would you say the root cause of that is that ::ChainstateActive is global?

Also, `chainActive`, `mapBlockIndex`, .... I fail to see how an interface could help here. Shouldn't `CChainState` be the interface?

**Comment by jamesob** [comment]
> An alternative could be to remove the global methods (e.g. ::ActivateBestChain) and inline them into the members?

I like this idea but left it on the table in the interest of minimizing the diff.

In general, my preference would be to have anything outside of validation deal with the CChainState interface (`::ChainstateActive()` or other instances), and have anything only to be used by validation in a static function outside of the header file. I didn't initially do this (to keep the diff small), but I'm happy to add commits if that seems preferable.

File: src/validation.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/validation.h]: in commit de39209d0c move-only: make the CChainState interface public

Could keep this implementation in the cpp file? Only put the struct and declaration in the header?

Also, the commit description mentions a "namespace", which is no longer true, so should be removed?

File: src/validation.h [hunk 3]
```diff
@@ -524,6 +533,21 @@ class CChainState {
...
+    /**
+     * Update the on-disk chain state.
+     * The caches and indexes are flushed depending on the mode we're called with
+     * if they're too large, if it's been a while since the last write,
+     * or always and in all cases if we're in prune mode and are deleting files.
+     *
+     * If FlushStateMode::NONE is used, then FlushStateToDisk(...) won't do anything
+     * besides checking if we need to prune.
+     */
+    bool FlushStateToDisk(
```
Comment (MarcoFalke) [inline src/validation.h]: In commit 9147f28ec0 refactoring: FlushStateToDisk -> CChainState


It seems weird to move a method that was `static` in validation.cpp to the header file as a public member function. At the very least it should be private. (However, I'd prefer if in the same commit the `FlushStateToDisk` and `PruneAndFlush` were moved to a public method in CChainState.)

File: src/validation.h [hunk 4]
```diff
@@ -523,6 +521,14 @@ class CChainState {
      */
     CCriticalSection m_cs_chainstate;
 
+    /**
+     * Is this chainstate undergoing initial block download?
```
Comment (MarcoFalke) [inline src/validation.h]: in commit 61fd34e244  refactoring: IsInitialBlockDownload -> CChainState


Instead of a question, say "Whether this c..."?

File: src/validation.h [hunk 5]
```diff
@@ -523,6 +521,14 @@ class CChainState {
...
+     *
+     * Mutable because we need to be able to mark IsInitialBlockDownload()
+     * const, which toggles this for caching purposes.
```
Comment (MarcoFalke) [inline src/validation.h]: in commit 61fd34e refactoring: IsInitialBlockDownload -> CChainState

This is only toggled once (aka latched)? So should keep that nomenclature?

**Review by maflcko — COMMENTED** [review]
utACK 61fd34e244

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

utACK 61fd34e244
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUihkgv/aDDUmSAn14AENjmciWELRSosOaTj/t0ws0/y09T+/f8wcmv/kCBx6tnR
JJkdcyteDAIlxJADIzT3TnhhXg93zwozotbv9hYGt/wTij4bLJIO96mBIWqJc/9V
ivfScok+S1CwLVRp8nT3mzJMWMQLFFp7CyfZmKFnrCWCI6tuK2GSqCuy5dKkDjgk
1hPhYTjk1OsYUhn2OjFVyfK4C1I6U8qfTFh6H6iSTAX4lA5/x9Jx/yINw93PDaH2
mMX8II9Pns7YY+zEXfNhw23IqomDe6GBTMBFw+6VAKAiGV5mIQeoabRcEdeqDT+4
xn6CzXaID4ENm+tqDkiaFTrSP7JW3BIt7Bnz3+D1NbPZzRow2rEwaF9pH8/DPmVk
WWU7ZziPywF2Q3BTCPq1SrkYgBSDYHYlA98XyZ4OAdhiT/tPV1gM5kihyy2LufpE
lec41niuCaOn8zDJeUF4KkhtVfMTdXE8f7od+i6ziBpCJO4UZ6fxxZWQFiDJQIJQ
1DFdLFPI
=enPL
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `2470d8f5b5b09630c89a5fc8a331144aed1939344367664ad30f7e38bc00a911  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401082470d8f5b5b09630c89a5fc8a331144aed1939344367664ad30f7e38bc00a911f010703373bd677c5e7f8f26fcd1aa1a8dc508fff010011f3b43fe4b4e5a67e9e62c4fde8db408f1045cd9ba31f008adbd7085894a220a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01002e8e20304bb407b6c568fcabbdca6ce08f1045cd9ba30f00823b2a1265b2361380083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0102f6dfe73f4a1b833a9417baf0e0d51ee08f01011359652058eed55b84c6ea13af079c808f1045cd9ba31f008938bc8c96ed06f2f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0103eb93d1760c27c8ace1353cc81f5100e08f1045cd9ba30f008120b2835f88c82200083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Review by ryanofsky — APPROVED** [review]
utACK 61fd34e24400b4cee07a379630331f37d2847588. Change is simple and looks good, though it might be nice to follow up on some of Marco's suggestions.

File: src/validation.h [hunk 6]
```diff
@@ -278,9 +279,9 @@ void PruneOneBlockFile(const int fileNumber) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 
 /** Flush all state, indexes and buffers to disk. */
-void FlushStateToDisk();
+void FlushStateToDisk(CChainState& chainstate);
 /** Prune block files and flush state to disk. */
-void PruneAndFlush();
+void PruneAndFlush(CChainState& chainstate);
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/validation.h]: No good reason beyond the fact that `FlushStateToDisk` affects on-disk block files (which are shared across chainstates) - but that's a symptom of a more general issue (that FSTD handles pruning, which seems like a separate concern) so probably not worth worrying about here.

File: src/validation.h (hunk 3 above)
Comment (jamesob) [reply src/validation.h]: Agree with your general concern, but usage of `FlushStateToDisk` happens in places that won't work with `private` like `AcceptToMemoryPoolWithTime` and so if we made it private, we'd have to attach a bunch of `friend` statements which seems kind of gross.

**Comment by jamesob** [comment]
Thanks for the good feedback @Sjors @MarcoFalke @ryanofsky. I agree with the comments so far and have incorporated all of them aside from those noted above. Initially I hadn't touched the global `Flush` functions because I thought it'd create an unnecessarily large diff, but it turns out that it's negligible.

I took the liberty of renaming the global function equivalents for clarity, and in the process I did notice something sort of interesting. `PruneAndFlush()` doesn't seem to actually do any flushing, just pruning, and so I'm calling it `PruneBlockFiles()` accordingly. This might be worth more general investigation because other parts of the code seem to assume that call flushes chainstate (which doesn't seem correct).

File: src/validation.h (hunk 3 above)
Comment (MarcoFalke) [reply src/validation.h]: Hmm, makes sense

File: src/validation.cpp [hunk 7]
```diff
@@ -4916,3 +4760,21 @@ class CMainCleanup
         mapBlockIndex.clear();
     }
 } instance_of_cmaincleanup;
+
+bool CBlockIndexWorkComparator::operator()(const CBlockIndex *pa, const CBlockIndex *pb) const {
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit a8930a66b1 move-only: make the CChainState interface public


Why is this moved to the end of the cpp file instead of keeping it at the top? This breaks both `git blame` and git diff with `--color-moved=dimmed-zebra --ignore-all-space`

File: src/validation.h [hunk 8]
```diff
@@ -435,6 +610,9 @@ bool InvalidateBlock(CValidationState& state, const CChainParams& chainparams, C
 /** Remove invalidity status from a block and its descendants. */
 void ResetBlockFailureFlags(CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+/** @returns the most-work chainstate. */
```
Comment (MarcoFalke) [inline src/validation.h]:
```suggestion
/** @returns the most-work valid chainstate. */
```

in commit 7794d41a58 refactoring: introduce ChainstateActive()


Could clarify that this the chainstate is fully validated (as opposed to a pow header chain)

File: src/validation.cpp [hunk 9]
```diff
@@ -2196,27 +2035,31 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
...
     }
     return true;
 }
 
-void FlushStateToDisk() {
+void CChainState::ForceFlushStateToDisk() {
     CValidationState state;
     const CChainParams& chainparams = Params();
-    if (!FlushStateToDisk(chainparams, state, FlushStateMode::ALWAYS)) {
+    if (!this->FlushStateToDisk(chainparams, state, FlushStateMode::ALWAYS)) {
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit 7ab9e339a8 refactoring: FlushStateToDisk -> CChainState


style-question: You remove the global with the same name in this commit, so the `this->` is not required?

File: src/validation.cpp [hunk 10]
```diff
@@ -2196,27 +2035,31 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
...
-void PruneAndFlush() {
+void CChainState::PruneBlockFiles() {
     CValidationState state;
     fCheckForPruning = true;
     const CChainParams& chainparams = Params();
-    if (!FlushStateToDisk(chainparams, state, FlushStateMode::NONE)) {
+
+    // Confusingly, nothing is actually flushed to disk due to
+    // FlushStateMode::NONE; the only effect of this call is to prune
+    // block files if necessary.
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit 7ab9e339a8 refactoring: FlushStateToDisk -> CChainState


I'd rather not add a comment that starts with "confusingly". You renamed the function and removed the confusion.

**Review by maflcko — COMMENTED** [review]
utACK 5fe4eb3a1a

File: src/validation.cpp (hunk 9 above)
Comment (jamesob) [reply src/validation.cpp]: I like prefixing with `this->` for methods because it makes it obvious we're calling a method, sort of like `m_`. See the IRC discussion: http://www.erisian.com.au/bitcoin-core-dev/log-2019-05-06.html#l-572

Happy to undo it if it's controversial though.

File: src/validation.cpp (hunk 7 above)
Comment (jamesob) [reply src/validation.cpp]: It'd be in an awkward place at the top of the file - right below the "global state" comment. Happy to move it back if you think the diff convenience is worth it though.

File: src/validation.h (hunk 8 above)
Comment (jamesob) [reply src/validation.h]: Ah good point! Thanks.

**Comment by jamesob** [comment]
> PruneAndFlush() doesn't seem to actually do any flushing

My mistake: flushes *do* happen when calling `PruneAndFlush()`. I've undone that particular rename and removed the comment I added.

File: src/validation.cpp (hunk 7 above)
Comment (MarcoFalke) [reply src/validation.cpp]: I don't understand what the "global state" comment should convey. If a symbol is global and not const, we prefix it with `g_`. Might want to remove it?

File: src/validation.cpp (hunk 9 above)
Comment (MarcoFalke) [reply src/validation.cpp]: Asked because we don't do this elsewhere. You can pick whatever you prefer ofc

Comment (ryanofsky) [reply src/validation.cpp]: > I like prefixing with this-> for methods because it makes it obvious we're calling a method, sort of like m_

FWIW, I still think https://github.com/bitcoin/bitcoin/pull/14635 is a good solution for this that would improve readability in these cases (https://github.com/bitcoin/bitcoin/pull/15976#issuecomment-490557315, https://github.com/bitcoin/bitcoin/pull/15976#issuecomment-490559432, https://github.com/bitcoin/bitcoin/pull/15976#discussion_r284264160)

**Review by ryanofsky — APPROVED** [review]
utACK 64d0f872fcf6b61de9e4e62ce13dbbe3b00cbb6c. Since previous: ForceFlushStateToDisk and PruneAndFlush are now class members, CBlockIndexWorkComparator compare method is moved to the cpp file, some comments and commit messages are updated

File: src/validation.cpp [hunk 11]
```diff
@@ -63,162 +63,10 @@
 /**
  * Global state
```
Comment (MarcoFalke) [inline src/validation.cpp]: nit, in commit a8930a66b1 move-only: make the CChainState interface public


This comment wasn't accurate before the commit, nor is it accurate after the commit. And it made you move the code block below all the way down in the file.

Could instead remove this comment and keep the code block below in place? This also makes `git blame` and `git diff` work on the commit without workarounds.

Comment (jamesob) [reply src/validation.cpp]: Yep, agree. Not sure how much it helps the diff, but that comment should be removed. I've pushed [the change](https://github.com/bitcoin/bitcoin/compare/64d0f872fcf6b61de9e4e62ce13dbbe3b00cbb6c..403e677c9ebbf9744733010e6b0c2d1b182ee850).

File: src/validation.h [hunk 12]
```diff
@@ -508,6 +513,28 @@ class CChainState {
...
+    bool FlushStateToDisk(
+        const CChainParams& chainparams,
+        CValidationState &state,
+        FlushStateMode mode,
+        int nManualPruneHeight = 0);
+
+    //! Unconditionally flush all changes to disk.
+    void ForceFlushStateToDisk();
+
+    //! Prune blockfiles from the disk if necessary and then flush chainstate changes
```
Comment (Sjors) [inline src/validation.h]: Nit: flushing is done _before_ pruning. Rename to `FlushAndPrune()`?

File: src/validation.cpp [hunk 13]
```diff
@@ -2058,27 +2046,28 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
...
-    if (!FlushStateToDisk(chainparams, state, FlushStateMode::ALWAYS)) {
+    if (!this->FlushStateToDisk(chainparams, state, FlushStateMode::ALWAYS)) {
         LogPrintf("%s: failed to flush state (%s)\n", __func__, FormatStateMessage(state));
     }
 }
 
-void PruneAndFlush() {
+void CChainState::PruneAndFlush() {
     CValidationState state;
     fCheckForPruning = true;
```
Comment (Sjors) [inline src/validation.cpp]: Yuck (but out of scope)

**Review by Sjors — APPROVED** [review]
utACK 403e677

File: src/validation.h (hunk 12 above)
Comment (jamesob) [reply src/validation.h]: Ehh, well it depends on what you're talking about being flushed - i.e. `pcoinsTip` get flushed [after](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2191) the prune happens.

Comment (Sjors) [reply src/validation.h]: I was talking about the chain state, which is the biggest thing :-)

Comment (jamesob) [reply src/validation.h]: Yeah, `pcoinsTip` (chainstate) is flushed *after* pruning.

Comment (Sjors) [reply src/validation.h]: Ah wait, never mind, you're right. Why do we flush the chainstate cache after pruning? Is that in order to minimize the chances of running out of disk space?

**Review by ryanofsky — APPROVED** [review]
utACK 403e677c9ebbf9744733010e6b0c2d1b182ee850. Only change since previous review is removing global state comment as suggested.

File: src/validation.cpp [hunk 14]
```diff
@@ -4414,7 +4269,7 @@ bool CChainState::LoadGenesisBlock(const CChainParams& chainparams)
 
 bool LoadGenesisBlock(const CChainParams& chainparams)
 {
-    return g_chainstate.LoadGenesisBlock(chainparams);
+    return ::ChainstateActive().LoadGenesisBlock(chainparams);
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit 4d6688603b refactoring: introduce ChainstateActive()


Why is this changed? It seems this method will go away anyway. See https://github.com/bitcoin/bitcoin/pull/15606/files#diff-24efdb00bfbe56b140fb006b562cc70bL4415

File: src/validation.cpp [hunk 15]
```diff
@@ -4312,7 +4167,7 @@ bool CChainState::RewindBlockIndex(const CChainParams& params)
 }
 
 bool RewindBlockIndex(const CChainParams& params) {
-    if (!g_chainstate.RewindBlockIndex(params)) {
+    if (!::ChainstateActive().RewindBlockIndex(params)) {
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit 4d66886 refactoring: introduce ChainstateActive()


Why is this changed? It seems that eventually you'll pass in the chainstate, see https://github.com/bitcoin/bitcoin/pull/15606/files#diff-24efdb00bfbe56b140fb006b562cc70bR4324

So why would you change the same line twice?

File: src/validation.cpp [hunk 16]
```diff
@@ -4179,7 +4034,7 @@ bool CChainState::ReplayBlocks(const CChainParams& params, CCoinsView* view)
 }
 
 bool ReplayBlocks(const CChainParams& params, CCoinsView* view) {
-    return g_chainstate.ReplayBlocks(params, view);
+    return ::ChainstateActive().ReplayBlocks(params, view);
```
Comment (MarcoFalke) [inline src/validation.cpp]: in commit 4d66886 refactoring: introduce ChainstateActive()

Same here (and all other changes in this commit)

File: src/validation.cpp [hunk 17]
```diff
@@ -60,165 +60,27 @@
...
 #define MILLI 0.001
 
-/**
- * Global state
- */
-namespace {
-    struct CBlockIndexWorkComparator
-    {
-        bool operator()(const CBlockIndex *pa, const CBlockIndex *pb) const {
-            // First sort by most total work, ...
```
Comment (MarcoFalke) [inline src/validation.cpp]: git-style-nit in commit d7c97edeea move-only: make the CChainState interface public


Could keep the previous indentation to preserve the `git blame` and minimize the diff line count? No strong opinion, though.
Please don't address my nit if you don't have to touch the patches for other reasons.

**Review by maflcko — COMMENTED** [review]
utACK 403e677c9e, though the diff still seems a bit bloated with some unnecessary changes in the second commit.

File: src/validation.cpp (hunk 17 above)
Comment (jamesob) [reply src/validation.cpp]: I don't know that we should maintain a weird indentation just for ease-of-diff, but if other reviewers disagree, I'll make the change.

File: src/validation.cpp (hunk 15 above)
Comment (jamesob) [reply src/validation.cpp]: Even before that change, we remove the `g_chainstate` symbol (https://github.com/bitcoin/bitcoin/pull/15606/commits/4c4b9ce2f310a519f1579f6a2892c590c1b96c6a) which means that this change would have to happen at some point anyway.

**Comment by MarcoFalke** [comment]
My comments are mostly meant to minimize the changed lines, so purely stylistic. I guess you shouldn't address them unless you need to invalidate the reviews for other reasons.

File: src/validation.h [hunk 18]
```diff
@@ -422,6 +423,145 @@ inline CBlockIndex* LookupBlockIndex(const uint256& hash)
...
+    DISCONNECT_UNCLEAN, // Rolled back, but UTXO set was inconsistent with block.
+    DISCONNECT_FAILED   // Something else went wrong.
+};
+
+class ConnectTrace;
+
+struct CBlockIndexWorkComparator
+{
+    bool operator()(const CBlockIndex *pa, const CBlockIndex *pb) const;
+};
```
Comment (Empact) [inline src/validation.h]: nit in commit d7c97ed move-only: make the CChainState interface public

How about defining `CBlockIndexWorkComparator` privately within `CChainState`, given it is not used outside the class?

File: src/validation.cpp [hunk 19]
```diff
@@ -80,6 +80,8 @@ bool CBlockIndexWorkComparator::operator()(const CBlockIndex *pa, const CBlockIn
 
 CChainState g_chainstate;
 
+CChainState& ChainstateActive() { return g_chainstate; }
```
Comment (Empact) [inline src/validation.cpp]: in commit 4d66886 refactoring: introduce ChainstateActive()

How about making `g_chainstate` static?

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15976/commits/403e677c9ebbf9744733010e6b0c2d1b182ee850 no need to address my nits herein

File: src/validation.h [hunk 20]
```diff
@@ -422,6 +418,186 @@ inline CBlockIndex* LookupBlockIndex(const uint256& hash)
...
+     * besides checking if we need to prune.
+     */
+    bool FlushStateToDisk(
+        const CChainParams& chainparams,
+        CValidationState &state,
+        FlushStateMode mode,
+        int nManualPruneHeight = 0);
+
+    //! Unconditionally flush all changes to disk.
+    void ForceFlushStateToDisk();
```
Comment (Empact) [inline src/validation.h]: nit: given context, this could be renamed `ForceFlushToDisk`

File: src/validation.h [hunk 21]
```diff
@@ -422,6 +418,186 @@ inline CBlockIndex* LookupBlockIndex(const uint256& hash)
...
+    /**
+     * Update the on-disk chain state.
+     * The caches and indexes are flushed depending on the mode we're called with
+     * if they're too large, if it's been a while since the last write,
+     * or always and in all cases if we're in prune mode and are deleting files.
+     *
+     * If FlushStateMode::NONE is used, then FlushStateToDisk(...) won't do anything
+     * besides checking if we need to prune.
+     */
+    bool FlushStateToDisk(
```
Comment (Empact) [inline src/validation.h]: nit: given context, this could be renamed `FlushToDisk`

File: src/validation.h [hunk 22]
```diff
@@ -422,6 +418,186 @@ inline CBlockIndex* LookupBlockIndex(const uint256& hash)
...
+
+class ConnectTrace;
+
+/** @see CChainState::FlushStateToDisk */
+enum class FlushStateMode {
+    NONE,
+    IF_NEEDED,
+    PERIODIC,
+    ALWAYS
+};
```
Comment (Empact) [inline src/validation.h]: nit: this could be moved to `CChainState::FlushMode`

**Comment by jamesob** [comment]
4 utACKs on this - is there anyone else who wants to give it a look before merge?

File: src/validation.h [hunk 23]
```diff
@@ -508,6 +513,28 @@ class CChainState {
 
     bool LoadBlockIndex(const Consensus::Params& consensus_params, CBlockTreeDB& blocktree) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
+    /**
+     * Update the on-disk chain state.
+     * The caches and indexes are flushed depending on the mode we're called with
```
[the remark this answers is not in the document]
Comment (jamesob) [reply src/validation.h]: I very much like the idea of clarifying this comment, but in my experience doc changes sometimes generate a surprising amount of feedback. I don't want to expand this PR beyond its current size (or delay it more than necessary), but will definitely be happy to review follow-up PRs making the change you suggest.

File: src/validation.cpp (hunk 17 above)
Comment (promag) [reply src/validation.cpp]: I don't mind the indentation `git show --color-moved=dimmed_zebra --color-moved-ws=ignore-all-space d7c97edeea8cee10ad9da1f940d39d5073ac142d`.

I think this operator was inlined? Does it make sense to declare it `inline`?

**Review by promag — COMMENTED** [review]
utACK 403e677 and rebased with current [master](c7cfd20a7).

**Comment by laanwj** [comment]
utACK 403e677c9ebbf9744733010e6b0c2d1b182ee850
